hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java

```diff
@@ -16,35 +16,553 @@
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
 
 /**
- * Keeps track of the amount of usage of different throttle categories (by {@code id}), and returns
- * whether the throttle has been exceeded after applying the given incremental amount.
+ * Keeps track of the amount of usage of different TPS throttle categories and gas, and returns whether a given
+ * transaction or query should be throttled based on that.
+ * Meant to be used in single-threaded context only as part of the {@link com.hedera.node.app.workflows.handle.HandleWorkflow}.
  */
-public interface ThrottleAccumulator {
+@Singleton
+public class ThrottleAccumulator {
+
+    private static final Logger log = LogManager.getLogger(ThrottleAccumulator.class);
+    private static final Set<HederaFunctionality> GAS_THROTTLED_FUNCTIONS =
+            EnumSet.of(CONTRACT_CALL_LOCAL, CONTRACT_CALL, CONTRACT_CREATE, ETHEREUM_TRANSACTION);
+    private final IntSupplier capacitySplitSource;
+    private final ConfigProvider configProvider;
+    private EnumMap<HederaFunctionality, ThrottleReqsManager> functionReqs = new EnumMap<>(HederaFunctionality.class);
+    private static final int UNKNOWN_NUM_IMPLICIT_CREATIONS = -1;
+    private boolean lastTxnWasGasThrottled;
+    private GasLimitDeterministicThrottle gasThrottle;
+    private List<DeterministicThrottle> activeThrottles = Collections.emptyList();
+
+    public ThrottleAccumulator(
+            @NonNull final IntSupplier capacitySplitSource, @NonNull final ConfigProvider configProvider) {
+        this.configProvider = requireNonNull(configProvider, "configProvider must not be null");
+        this.capacitySplitSource = capacitySplitSource;
+    }
+
+    /*
+     * Updates the throttle requirements for the given transaction and returns whether the transaction should be throttled.
+     *
+     * @param txnInfo the transaction to update the throttle requirements for
+     * @param consensusTime the consensus time of the transaction
+     * @param state the current state of the node
+     * @return whether the transaction should be throttled
+     */
+    public boolean shouldThrottle(
+            @NonNull final TransactionInfo txnInfo,
+            @NonNull final Instant consensusTime,
+            @NonNull final HederaState state) {
+        resetLastAllowedUse();
+        lastTxnWasGasThrottled = false;
+        if (shouldThrottleTxn(txnInfo, consensusTime, state)) {
+            reclaimLastAllowedUse();
+            return true;
+        }
+
+        return false;
+    }
+
+    /*
+     * Updates the throttle requirements for the given query and returns whether the query should be throttled.
+     *
+     * @param queryFunction the functionality of the query
+     * @param now the time at which the query is being processed
+     * @param query the query to update the throttle requirements for
+     * @return whether the query should be throttled
+     */
+    public boolean shouldThrottle(
```

Comment:
⚠️ I think we should add a parameter `AccountID queryPayerId` to this method's arguments and then do the same `isPayerThrottleExempt()` check you do [here in the transaction throttling logic](https://github.com/hashgraph/hedera-services/blob/08355-implement-frontend-throttle/hedera-node/hedera-app/src/main/java/com/hedera/node/app/throttle/ThrottleAccumulator.java#L230); this should allow the `callLocalDoesNotCheckSignaturesNorPayer()` spec below to continue passing.